server/aesdsocket.c

```diff
@@ -15,42 +15,245 @@
 #include <arpa/inet.h>
 #include <fcntl.h>
 #include <sys/types.h>
+#include <stdbool.h>
+#include <pthread.h>
+#include <sys/time.h>
+#include "queue.h"
+#include <errno.h>
 
 #define PORT 9000
 #define DATA_FILE "/var/tmp/aesdsocketdata"
 #define BUFFER_SIZE 1024
-#define ERROR_MEMORY_ALLOC   -6
+#define ERROR_MEMORY_ALLOC -6
 #define FILE_PERMISSIONS 0644
 
 int server_socket;
 int client_socket;
 int daemon_mode = 0;
 int option_value = 1;
 
+int file_fd = 0;
+// int data_count = 0;	
+
+bool sig_handler_hit = false;
+bool timer_thread_flag = false;
+
+pthread_t timer_thread = (pthread_t)NULL;
+
+void *thread_func(void *thread_parameter);
+
 // Function to handle signals
 static void signal_handler(int signo);
 
 // Function to daemonize the process
 void daemonize();
 
+void socket_connect(void);
+
+void exit_safely();
+
 int main(int argc, char **argv);
 
+//  Thread parameter structure
+struct thread_data {
+	pthread_t thread_id; // ID returned by pthread_create()
+	int client_socket; // holds the current client socket identifier
+
+    /**
+     * Set to true if the thread completed with success, false
+     * if an error occurred.
+     */
+	bool thread_complete;
+};
+
+// SLIST.
+typedef struct slist_data_s slist_data_t;
+struct slist_data_s {
+	struct thread_data connection_data_node;
+	SLIST_ENTRY(slist_data_s) entries;
+};
+
+slist_data_t *datap = NULL;
+
+SLIST_HEAD(slisthead, slist_data_s) head;
+// SLIST_INIT(&head);
+
+pthread_mutex_t mutex_lock = PTHREAD_MUTEX_INITIALIZER;
+
+
+void exit_safely() {
+	shutdown(server_socket, SHUT_RDWR);
+	unlink(DATA_FILE);
+    close(server_socket); // Close the server socket
+    close(client_socket); // Close the client socket
+    closelog(); // Close syslog
+
+    remove(DATA_FILE); // Remove the data file
+    exit(0); // Exit the program
```

Comment:
@sane5805 missing cleanup for threads which didn't exit in the main loop, however it's not safe to add in signal handler.  You need to re-think the cleanup strategy to do this safely.